react/src/topk/Topk.tsx

```diff
@@ -0,0 +1,449 @@
+  tokens,
+  activations,
+  firstDimensionName = "Sample",
+  secondDimensionName = "Layer",
+  thirdDimensionName = "Neuron" // Note that we simply use neuron as variable names throughout this file
+}: TopkProps) {
+  const activationsTensors = activations.map((sampleActivations) => {
+    return tensor<Rank.R3>(sampleActivations);
+  });
+  // Get number of layers/neurons (Assumes all samples have the same number of layers/neurons)
+  const numberOfLayers = activationsTensors[0].shape[1];
+  const numberOfNeurons = activationsTensors[0].shape[2];
+  const numberOfSamples = activationsTensors.length;
+
+  /** TODO: reqct-hook-form <- investigate */
+  const [sampleNumber, setSampleNumber] = useState<number>(0);
+  const [layerNumber, setLayerNumber] = useState<number>(0);
+  const [colsToShow, setColsToShow] = useState<number>(5);
+  const [k, setK] = useState<number>(Math.min(5, tokens[0].length));
+  const [neuronNumbers, setNeuronNumbers] = useState<number[]>([
+    ...Array(colsToShow).keys()
+  ]);
+  // Filter for whether to show the topk, bottomk or both (written as "bottomk+topk")
+  const [filter, setFilter] = useState<string>("topk+bottomk");
+
+  useEffect(() => {
+    // When the user changes the colsToShow, update the neuronNumbers
+    setNeuronNumbers(numberOfSamples > 1 ? [...Array(colsToShow).keys()] : [0]);
+  }, [colsToShow, numberOfSamples]);
+
+  const currentTokens: string[] = tokens[sampleNumber];
+  // Get the relevant activations for the selected sample, layer, and neurons.
+  const currentActivations: Tensor2D = getSelectedActivations(
+    activationsTensors[sampleNumber],
+    layerNumber,
+    neuronNumbers[0],
+    neuronNumbers[neuronNumbers.length - 1]
+  ); // [neurons x tokens]
+
+  const { values: topkValsRaw, indices: topkIdxsRaw } = tfTopk(
+    currentActivations,
+    k,
+    true
+  );
+  const { values: bottomkValsRaw, indices: bottomkIdxsRaw } = tfTopk(
+    currentActivations.mul(-1),
+    k,
+    true
+  );
+
+  // The topk and bottomk values, indices and tokens will be tensors of shape
+  // [tokens x neurons]. This form makes it easier to display the table
+  const topkVals: number[][] = topkValsRaw
+    .transpose()
+    .arraySync() as number[][];
+  const topkIdxs: number[][] = topkIdxsRaw
+    .transpose()
+    .arraySync() as number[][];
+  // Bottom vals are ordered from highest to lowest activations (just like top vals)
+  const bottomkVals: number[][] = reverse(bottomkValsRaw.mul(-1), -1)
+    .transpose()
+    .arraySync() as number[][];
+  const bottomkIdxs: number[][] = reverse(bottomkIdxsRaw, -1)
+    .transpose()
+    .arraySync() as number[][];
+
+  const topkTokens: string[][] = topkIdxs.map((outerArr) =>
+    outerArr.map((token_idx) => currentTokens[token_idx])
+  );
+  const bottomkTokens: string[][] = bottomkIdxs.map((outerArr) =>
+    outerArr.map((token_idx) => currentTokens[token_idx])
+  );
+
+  const selectRowStyle = {
+    paddingTop: 5,
+    paddingBottom: 5
+  };
```

Comment:
For future reference:

- It would have been nicer to extract this as a helper function so that it can be more easily tested.
- https://www.npmjs.com/package/react-hook-form is worth learning

Reply:
By "this" do you mean the whole code block? I.e. a function which returns the ~20 variables?